### PR TITLE
Adding tolerances when finding AuxDet(Sensitive)s

### PR DESCRIPTION
**Pull request by ilepetic**
Changes to FindAuxDetSensitiveAtPosition and related functions to allow for a tolerance in finding matching AuxDets or AuxDetSensitives. The tolerance defaults to 0.

### CODE REVIEW

File: larcorealg/Geometry/GeometryCore.h [hunk 1]
```diff
@@ -4573,93 +4573,104 @@ namespace geo {
     /**
      * @brief Returns the index of the auxiliary detector at specified location.
      * @param worldLoc 3D coordinates of the point (world reference frame)
+     * @param tolerance tolerance for matches. Default 0
```
Comment (knoepfel) [inline larcorealg/Geometry/GeometryCore.h]: Here and elsewhere, you need to document the units of the tolerance parameter (is it a relative tolerance, absolute in units of `cm`, etc.).

File: larcorealg/Geometry/AuxDetChannelMapAlg.cxx [hunk 2]
```diff
@@ -69,13 +70,14 @@ namespace geo{
...
+      if(localPoint[2] >= - (adsg.Length()/2 + tolerance)      &&
+         localPoint[2] <=   adsg.Length()/2 + tolerance      &&
+         localPoint[1] >= - (adsg.HalfHeight() + tolerance)    &&
+         localPoint[1] <=   adsg.HalfHeight() + tolerance    &&
          // if AuxDet a is a box, then HalfSmallWidth = HalfWidth
-         localPoint[0] >= - HalfCenterWidth + localPoint[2]*(HalfCenterWidth - adsg.HalfWidth2())/(0.5 * adsg.Length()) &&
-         localPoint[0] <=   HalfCenterWidth - localPoint[2]*(HalfCenterWidth - adsg.HalfWidth2())/(0.5 * adsg.Length())
+         localPoint[0] >= - HalfCenterWidth + localPoint[2]*(HalfCenterWidth - adsg.HalfWidth2())/(0.5 * adsg.Length()) - tolerance &&
+         localPoint[0] <=   HalfCenterWidth - localPoint[2]*(HalfCenterWidth - adsg.HalfWidth2())/(0.5 * adsg.Length()) + tolerance
          ) return a;
```
Comment (knoepfel) [inline larcorealg/Geometry/AuxDetChannelMapAlg.cxx:81]: This pattern is repeated enough that it is worth having a data structure that supports this kind of comparison.  I think a clearer coding pattern would be something like:

```c++
namespace lar {
  class BoxBoundaries {
    double xhw_, yhw_, zhw_, tolerance_;
  public:
    Box(double xhw, double y_hw, double zhw, double tolerance = 0.)
      : xhw_{std::abs(xhw)}, yhw_{std::abs(yhw)}, zhw_{std::abs(zhw)}, tolerance_{tolerance}
    {}

    bool contains(double const* point) const
    {
      auto x = point[0];
      auto y = point[1];
      auto z = point[2];
      if (std::abs(x) > xhw_ + tolerance_) return false;
      if (std::abs(y) > yhw_ + tolerance_) return false;
      if (std::abs(z) > zhw_ + tolerance_) return false;
      return true;
    }
  };
}

// ...
BoxBoundaries box{x_halflength, y_halflength, z_halflength, tolerance};
if (box.contains(localPoint)) return a;
```
Of course, you may need to handle a box center which is different than (0,0,0), but I hope this gives you the general idea.  This should result in easier to read code, which is harder to get wrong.

**Review by knoepfel — COMMENTED** [review]
Thanks for the PR, @ilepetic.  I recommend two things:

1. Document the units of the `tolerance` parameter in the Doxygen comments.
2. Encapsulate the box comparisons inside of a class.

I know you inherited the coding pattern for the latter--but now's a good time to clean it up.  Let me know if you have any questions.

File: larcorealg/Geometry/GeometryCore.h (hunk 1 above)
Comment (ilepetic) [reply larcorealg/Geometry/GeometryCore.h]: Units included

File: larcorealg/Geometry/AuxDetChannelMapAlg.cxx (hunk 2 above)
Comment (ilepetic) [reply larcorealg/Geometry/AuxDetChannelMapAlg.cxx:81]: Can we make this a feature request for a future release?

Comment (knoepfel) [reply larcorealg/Geometry/AuxDetChannelMapAlg.cxx:81]: I'm okay with that.

**Comment by lgarren** [comment]
@ilepetic still waiting for the changes on this PR

**Comment by wesketchum** [comment]
To avoid issues with building in uboonecode and icaruscode, I've removed the `tolerance` parameter from the `AuxDetChannelMapAlg::PositionToAuxDetChannel` function, which is pure virtual. In implementation for sbndcode, the tolerance is set and used there. I've checked that this builds locally with both an updated sbndcode branch and an icaruscode dev area (that has no changes in its CRTChannelMapAlg).

**Comment by knoepfel** [comment]
trigger build

**Comment by lgarren** [comment]
approve

**Pull request merged**